File: code/python/MinStack.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        self.minNum = 0

    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        if self.stack == []:
            self.stack.append(0)
            self.minNum = x
        else:
            self.stack.append(x-self.minNum)
            if x < self.minNum:
                self.minNum = x
        

    def pop(self):
        """
        :rtype: void
        """
        popped = self.stack.pop()
        if popped < 0:
            self.minNum -= popped

        

    def top(self):
        """
        :rtype: int
        """
        top = self.stack[-1]
        if top>0:
            return top + self.minNum
        else:
            return self.minNum
        

    def getMin(self):
        """
        :rtype: int
        """
        return self.minNum


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: code/java/minStack.java

class MinStack {

    /** initialize your data structure here. */
    Stack<Long> s;
    long min;
    public MinStack() {
        s = new Stack<>();
    }
    
    public void push(int x) {
        if(s.isEmpty()){
            s.push(0L);
            min = x;
            
        }
        else{
            s.push(x-min);
            if(x<min)
                min = (long)x;
        }
        
    }
    
    public void pop() {
        long pop = s.pop();
        if(pop<0)
            min = min - pop;
    }
    
    public int top() {
        long top = s.peek();
        if(top > 0)
            return (int)(min + top);
        else
            return (int)min;
    }
    
    public int getMin() {
        return (int)min;

    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */
